feat: Dynamic Menu Items - Menu Button Widget

dhruvikn opened the pull request:
First draft for #636

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/appsmithorg/appsmith-docs?pullRequest=649) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/appsmithorg/appsmith-docs?pullRequest=649) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/appsmithorg/appsmith-docs?pullRequest=649) it.</sub>

dhruvikn commented:
Create the PR against main instead of v1.4. Therefore, closing it.

dilippitchika commented on website/docs/reference/widgets/menu-button.md:
```diff
@@ -74,3 +74,72 @@ Each menu item can now be customized through a set of properties by clicking on
+
+With [#17652](https://github.com/appsmithorg/appsmith/pull/17652), users will be able to add upto 10 menu items using JS. They simply need to select Menu Item Source (a new property) as Dynamic, bind a source data, and then they can configure each menu item by referencing `currentItem` inside the configuration.
+
+### TL;DR
```
@dhruvikn rather than a TL;DR section. It's better if you add a loom video, so that the team can create a new video from it and use it.

dilippitchika reviewed (commented):
@dhruvikn can you add a screenshot of the error state when menu items exceed 10?

somangshu reviewed (approved):
LGTM
Agreed with Dilip's comment

dhruvikn commented:
I have added the screenshot of the error state.

Also, I have already added a video in the issue #636's description which the team can refer. In case you missed it, [it's here](https://www.youtube.com/watch?v=ajzBkooI7CE).

cc: @dilippitchika  @somangshu

jnikhila commented:
@dilippitchika, @dhruvikn  - The changes are for the `Menu Button` Widget. A new property - `Menu Items Source` is added. Could you please confirm the below understanding:

* Menu Items Source is a widget property.
* The previous implementation for menu item is now available as part of Menu Items Source - Static.
*  Menu Items Source - has two source types (Static) and a new Source 'Dynamic'.
* Once user selects Dynamic source, user can further set properties and styles.
* `onClick` can fetch the selected values by using currentItem.<attribute_name> of the object.

dhruvikn commented:
Yes @jnikhila, that's correct. Just a couple of pointers though - 

- Once user selects Dynamic source, user can further set properties and styles. **_They can also utilize currentItem and currentIndex here._**
- We have also added two more new widget properties - `Source Data` and `Configure Menu Items`. These properties are only required when `Menu Items Source` is set to `Dynamic`.

danciaclara reviewed (changes requested):
Left comments. We are also  missing a 'Events' subsection for this page

jnikhila commented on website/docs/reference/widgets/menu-button.md:
```diff
@@ -1,76 +1,107 @@
 # Menu Button
 
-Menu buttons are used to represent a set of actions in a group.
+The Appsmith menu buttons widget is a customizable dropdown menu that allows you to add a list of options for users to select from. It's a powerful tool for creating navigation menus, dropdown lists, and options for users to choose from.
```
Adding Appsmith helps in branded keyword searches in SEO. So kept it there.

danciaclara replied:
Hmm I understand that. But I think we should remove this as it looks a bit odd.

jnikhila replied:
Okay.

jnikhila commented on website/docs/reference/widgets/menu-button.md:
```diff
@@ -1,76 +1,107 @@
 # Menu Button
 
-Menu buttons are used to represent a set of actions in a group.
+The Appsmith menu buttons widget is a customizable dropdown menu that allows you to add a list of options for users to select from. It's a powerful tool for creating navigation menus, dropdown lists, and options for users to choose from.
```
I think this description helps all types of users visualize using the Menu button widget. The earlier sentence would only be understandable to advanced users who would know actions in Appsmith's reference.

danciaclara replied:
I don't think it is referring to our Framework actions. I think it is just referring to a group of actions in a dropdown menu. But yeah you are right that it can be confusing. So let's go with your version :)

jnikhila commented on website/docs/reference/widgets/menu-button.md:
```diff
@@ -1,76 +1,107 @@
 
-## Properties
+:::info
+The icons are used from [Blueprint](https://blueprintjs.com) library. For more information, see the [list of icons available on Blueprint](https://blueprintjs.com/docs/#icons).
```
Sorry! Didn't understand this. Could you please elaborate?

danciaclara replied:
I mean that instead of adding that line as a callout, let's add that statement where the red underline is on the screenshot cause it is relevant info there to let the users know which set of icons are available to them.

jnikhila replied:
Got it.

jnikhila commented on website/docs/reference/widgets/menu-button.md:
```diff
@@ -1,76 +1,107 @@
+| ------------------- | ---------------- |------------------- | ---------------- |
+| **Label**        | Formatting/ Binding | This property sets the label of the menu button. It can be a static text value or dynamically bound to a source of data. For example, you might set the label to be the logged-in user’s name. To dynamically bind the text, use the `{{}}` mustache sign. You can also read the value of the label in the code. | `{{<WIDGET_NAME.label}}`|
+| [**Menu Items Source**](#menu-items-source)| Formatting | This property allows you to specify the source for the menu items. You can define a static list of items or provide a dynamic data source to be displayed as menu items. |                   |
+| **Disabled**         | Formatting | This property disables the input to the widget. For more information, see [Disabled property details](/reference/widgets#disabled).|                  |
```
It is added to the section `##### Menu Items
This property allows you to define the options you can show in the dropdown list when the menu button is clicked. You can add the items from the property pane by using the `Add a New Menu Item` button. Each menu item can be customized using the gear icon available next to it. For example, you might define a menu item with a label "Home" and an action of "Navigate to Home page" by using the gear icon.`

jnikhila commented:
> Hey @danciaclara @jnikhila, just putting this here - I'll be happy to jump on a call to discuss the widget in depth and clarify any doubts you might have. Please let me know if needed. 🙂

Thanks @dhruvikn - Could you please share your thoughts on [this comment](https://github.com/appsmithorg/appsmith-docs/pull/649#discussion_r1045733598)?

dhruvikn commented:
Hey @jnikhila, I completely agree with @danciaclara's [comment](https://github.com/appsmithorg/appsmith-docs/pull/649#discussion_r1045733598). `Menu Item Properties` are the same for both `Menu Items Source` - `Static` and `Dynamic`. We can remove the `Dynamic Menu Item properties` section as it looks repeated.

Furthermore, cross-referencing, as @danciaclara suggested, 

> Just in the main Menu Button properties table, cross reference to the Static and Dynamic subsections in the description.

sounds great and helpful.

jnikhila commented:
> Hey @jnikhila, I completely agree with @danciaclara's [comment](https://github.com/appsmithorg/appsmith-docs/pull/649#discussion_r1045733598). `Menu Item Properties` are the same for both `Menu Items Source` - `Static` and `Dynamic`. We can remove the `Dynamic Menu Item properties` section as it looks repeated.
> 
> Furthermore, cross-referencing, as @danciaclara suggested,
> 
> > Just in the main Menu Button properties table, cross reference to the Static and Dynamic subsections in the description.
> 
> sounds great and helpful.

I thought @danciaclara is asking to remove the Item configuration properties. @danciaclara - Are you suggesting to remove the two tables that we added for static and dynamic items properties or should we keep one table properties and call it as Menu Item source properties and then cross-reference to static and dynamic. 
@dhruvikn - Your thoughts?

jnikhila commented on website/docs/reference/widgets/menu-button.md:
```diff
@@ -1,76 +1,107 @@
+|**Background color**| Style | This property sets the background color of the menu item. This can be useful for adding visual interest or highlighting the selected menu item.| |
 
-Style properties allow you to change the look and feel of the widget.
+### Dynamic Menu Item properties
```
Based on the [discussion](https://github.com/appsmithorg/appsmith-docs/pull/649#issuecomment-1347896001), I've changed the heading. Please review and let me know what you think.

dhruvikn commented:
@jnikhila - I think the changes that you pushed recently are exactly what I and @danciaclara were talking about. 🙂 
@danciaclara please confirm once.

However, one nitpick from me.
 
![CleanShot 2022-12-13 at 16 09 41@2x](https://user-images.githubusercontent.com/22471214/207298818-21f7681e-9718-47b5-909d-a54264f73f6a.png)

danciaclara commented on website/docs/reference/widgets/menu-button.md:
```diff
@@ -1,76 +1,89 @@
+| **Compact**         | Formatting | This property decides if the widget is in compact mode. In compact mode, the widget takes up less space on the page and can be suitable to use in tight layouts.|                 
 
-These properties allow you to edit the Menu Button widget. All these properties are present in the property pane of the widget. The following table lists all the widget properties.
+### Menu Items Source properties
```
Let this be Menu Item properties.  the 'Source' might be confusing for users

jnikhila replied:
Menu Item is only available when a user chooses static source. That's why Menu Items Source is added so that users know that it's applicable to both static as well as dynamic sources.

jnikhila commented on website/docs/reference/widgets/menu-button.md:
```diff
@@ -1,76 +1,89 @@
 
-These properties allow you to bind your widget with any other widget in queries or JS objects. The following table lists all the binding properties.
+## Menu Items Source
+The "Menu Items Source" property lets you define where the menu items come from. You can either define them manually (Static) or use a dynamic source by selecting the "Dynamic" option. Dynamic option allows you to specify a data source such as a [database query](/core-concepts/data-access-and-binding/querying-a-database) or [API](/core-concepts/connecting-to-data-sources/authentication/connect-to ...
```
@dilippitchika - `one callout here, even for static menu items, you can still link the label or any other property to come from a query/API.` is added as part of the label, where it is specified that it could be a static text or bind to a dynamic source.
<img width="870" alt="Screenshot 2022-12-14 at 5 25 49 PM" src="https://user-images.githubusercontent.com/99252011/207589213-4b3a59f2-8f00-49c7-87b2-03a99306cac6.png">

- For the Dynamic menu, the items are fixed - only 10 right?
- The items are created as per source data - Please refer to below screenshot - Do you think I should add that each item is not created individually?
<img width="886" alt="Screenshot 2022-12-14 at 5 27 41 PM" src="https://user-images.githubusercontent.com/99252011/207589662-f08c0f97-00cc-421e-8fe7-e62098de8a0a.png">

Please confirm

dilippitchika replied:
- Max 10 items, when dynamic source is selected
- Yes, the explanation in both label and dynamic is fine. 
- I was just calling it out in the menu items source section, don't know if we need to act on it

jnikhila replied:
@dilippitchika - Menu Items source - It serves as a main section - which briefs about the property and then has individual subsections to define it, so added to individual section. @danciaclara - Your thoughts?

jnikhila replied:
> * Max 10 items, when dynamic source is selected
> * Yes, the explanation in both label and dynamic is fine.
> * I was just calling it out in the menu items source section, don't know if we need to act on it

> * Max 10 items, when a dynamic source is selected - Yes, for dynamic source. Just a follow-up question -  If the source is static the user can add any number of menus, we don't have an upper bracket.

danciaclara replied:
I think it's ok to add it under Dynamic Source too. It's just a minor change in the last line. Let's modify it. Also not sure if we adding the part about max 10 items. Let's mention it if we haven't already.

danciaclara commented on website/docs/reference/widgets/menu-button.md:
```diff
@@ -1,76 +1,89 @@
-| ------------------ | ----------------------------------------- |
-| **Icon**           | Sets the icon to be used for menu button. |
-| **Icon alignment** | Sets the alignment of the selected icon.  |
+If the dynamic source array has more than 10 items, the `SourceData` fields show an error message: `Source data cannot have more than 10 items`.
```
this should be in the troubleshooting guide with the error message. no need to add it here, as we have already mentioned it in the callout. Is that ok @dilippitchika

danciaclara commented on website/docs/reference/widgets/menu-button.md:
```diff
@@ -1,76 +1,89 @@
-| **Visible**                        | Controls the visibility of a menu item. |
-| **Disabled**                       | Disables input to a menu item.          |
+:::note
+ Dynamic menu items are **only** supported inside the menu button and the list widget. You can add up to 10 menu items by defining them in the "Source Data" property for a dynamic source.
```
Is this line (Dynamic menu items are **only** supported inside the menu button and the list widget.) important here? We are talking about Dynamic property in the menu button page. so its fine not to mention this here.  I think we only need to mention about the max 10 menu items here.

jnikhila commented:
@danciaclara - Removed the error message and added ticket #683 to add it to the Troubleshooting section.
